File: 113-path-sum-ii/113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        ans = []
        
        def dfs(root, cur):
            if not root: return
            if not root.left and not root.right:
                if sum(cur) + root.val == targetSum:
                    ans.append(cur+[root.val])
                return
            dfs(root.left, cur+[root.val])
            dfs(root.right, cur+[root.val])
            
        dfs(root, [])
        return ans

File: 113-path-sum-ii/113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>>ans;
    void dfs(TreeNode* root, int target, vector<int> v) {
        if (!root)
            return;
        if (!root->left and !root->right and root->val == target){
            v.push_back(root->val);
            ans.push_back(v);
            return;
        }
        v.push_back(root->val);
        dfs(root->left, target-root->val, v);
        dfs(root->right, target-root->val, v);
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<int> cur;
        dfs(root, targetSum, cur );
        return ans;
    }
};